concent_api/core/payments/report.py

```diff
@@ -0,0 +1,214 @@
+    return queryset.select_related('payment_info')
+
+
+def get_related_payments_for_pair(clients_ethereum_addresses: list) -> Payments:
+    """ Get all payments between given requestor/provider pair. """
+    regular_payments = []  # type: List[str]
+    settlement_payments = []  # type: List[str]
+    forced_subtask_payments = []  # type: List[str]
+    requestor_additional_verification_payments = []  # type: List[str]
+    provider_additional_verification_payments = []  # type: List[str]
+    oldest_subtask_creation_date = get_report_computed_task_timestamp_of_oldest_subtask()
+    for clients_pair in clients_ethereum_addresses:
+        regular_payments = service.get_list_of_payments(  # pylint: disable=no-value-for-parameter
+            requestor_eth_address=clients_pair.requestor,
+            provider_eth_address=clients_pair.provider,
+            min_block_timestamp=oldest_subtask_creation_date,
+            transaction_type=TransactionType.BATCH,
+        )
+
+        settlement_payments = service.get_list_of_payments(  # pylint: disable=no-value-for-parameter
+            requestor_eth_address=clients_pair.requestor,
+            provider_eth_address=clients_pair.provider,
+            min_block_timestamp=oldest_subtask_creation_date,
+            transaction_type=TransactionType.SETTLEMENT,
+        )
+
+        forced_subtask_payments = service.get_list_of_payments(  # pylint: disable=no-value-for-parameter
+            requestor_eth_address=clients_pair.requestor,
+            provider_eth_address=clients_pair.provider,
+            min_block_timestamp=oldest_subtask_creation_date,
+            transaction_type=TransactionType.FORCED_SUBTASK_PAYMENT,
+        )
+
+        requestor_additional_verification_payments = service.get_covered_additional_verification_costs(  # pylint: disable=no-value-for-parameter
+            client_eth_address=clients_pair.requestor,
+            payment_ts=oldest_subtask_creation_date,
+        )
+        provider_additional_verification_payments = service.get_covered_additional_verification_costs(  # pylint: disable=no-value-for-parameter
+            client_eth_address=clients_pair.provider,
+            payment_ts=oldest_subtask_creation_date,
+        )
+
+    return Payments(
+        regular_payments=regular_payments,
+        settlement_payments=settlement_payments,
+        forced_subtask_payments=forced_subtask_payments,
+        requestor_additional_verification_payments=requestor_additional_verification_payments,
+        provider_additional_verification_payments=provider_additional_verification_payments,
+    )
+
+
+def get_report_computed_task_timestamp_of_oldest_subtask() -> int:
+    """
+    Return ReportComputedTask timestamp of oldest Subtask stored in database.
+    If there are no Subtasks in database, return current timestamp.
+    """
+    subtasks_set = Subtask.objects.order_by('created_at')
+    if subtasks_set.exists():
```

Comment:
There is no need to check this, it will actually do one more database query, it is enough to iterate over subtasks, and remove `else` below.

Reply:
Done